app/src/main/java/com/fcaputo/parkingapp/mvp/presenter/ReservationPresenter.kt

```diff
@@ -0,0 +1,80 @@
+package com.fcaputo.parkingapp.mvp.presenter
+
+import com.fcaputo.parkingapp.entity.Reservation
+import com.fcaputo.parkingapp.mvp.contract.ReservationContract
+import com.fcaputo.parkingapp.utils.Constants
+import com.fcaputo.parkingapp.utils.DateTimeCustomFormat
+import com.fcaputo.parkingapp.utils.validation.ValidationErrorType
+import java.time.LocalDateTime
+import java.time.LocalTime
+import java.util.Calendar
+import java.util.Date
+import java.util.TimeZone
+
+class ReservationPresenter(private val model: ReservationContract.Model, private val view: ReservationContract.View) : ReservationContract.Presenter {
+    init {
+        view.onSaveButton { onSaveButton() }
+        view.onDateInputPressed(this::onDateInputClick)
+        view.onTimeInputPressed(this::onTimeInputClick)
+        view.showParkingLotSize(model.getParkingLotSize())
+    }
+
+    override fun onDateInputClick(dateInputText: CharSequence?, editTextId: Int) {
+        //If a date has already been selected, return it, otherwise return today's date
+        val dateToShow = if (!dateInputText.isNullOrEmpty()) {
+            DateTimeCustomFormat.parseDate(dateInputText.toString())
+        } else { Date() }
+
+        view.showDatePicker(dateToShow, editTextId)
+    }
+
+    override fun onTimeInputClick(timeInputText: CharSequence?, editTextId: Int) {
+        //If a time has already been selected, return it, otherwise return now's time
+        val timeToShow = if (!timeInputText.isNullOrEmpty()) {
+            DateTimeCustomFormat.parseTime(timeInputText.toString())
+        } else { LocalTime.now() }
+
+        view.showTimePicker(timeToShow, editTextId)
+    }
+
+
+    override fun onSaveButton() {
+        if (formIsIncomplete()) {
+            view.showErrorMessage(ValidationErrorType.INCOMPLETE_FIELD)
+            return
+        }
+
+        val startDateTime = getDateTimeFromString(view.getStartDate(), view.getStartTime())
+        val endDateTime = getDateTimeFromString(view.getEndDate(), view.getEndTime())
+        val parkingLot = view.getParkingLotNumber().toInt()
+        val securityCode = view.getSecurityCode().toInt()
+
+        val validationResult = model.validateData(startDateTime, endDateTime, parkingLot)
+        if (validationResult.isSuccess) {
+            model.storeReservation(Reservation(startDateTime, endDateTime, parkingLot, securityCode))
+            view.showSuccessMessage()
+        } else {
+            validationResult.error?.let { view.showErrorMessage(it) }
+        }
+    }
+
+    private fun formIsIncomplete(): Boolean = view.getTextFieldsContents().any { it.isNullOrEmpty() }
```

Comment:
Suggestion: method names are more readable if they refer to positive actions, like `formIsComplete()`.

Reply:
Got it 👌

Reply:
Done